src/execution-chain/crosschain/ActionRelay.sol

```diff
@@ -64,53 +83,79 @@ contract ActionRelay is OAppSenderUpgradeable, UUPSUpgradeable {
 
     /// @notice Quote the messaging fee required to relay actions to another chain.
     /// @param _callId The unique identifier for the relayed actions, such as a proposal ID.
-    /// @param _actions The actions to relay to the destination chain, including value, target and calldata.
-    /// @param _allowFailureMap A bitmap of actions that are allowed to fail.
-    /// @param _dstEid The LayerZero endpoint ID of the destination chain.
     /// @param _gasLimit The additional gas needed on the destination chain to process the message, surplus will be refunded.
     function quote(
         uint256 _callId,
-        IDAO.Action[] memory _actions,
-        uint256 _allowFailureMap,
-        uint32 _dstEid,
         uint128 _gasLimit
     ) external view returns (LzSendParams memory params) {
-        bytes memory message = abi.encode(_callId, _actions, _allowFailureMap);
+        QueuedActionRelayParams memory action = actionsMap[_callId];
+
+        require(action.message.length > 0, "ActionRelay: no message to relay");
+
         bytes memory options = OptionsBuilder.newOptions().addExecutorLzReceiveOption({
             _gas: _gasLimit,
             _value: 0
         });
         MessagingFee memory fee = _quote({
-            _dstEid: _dstEid,
-            _message: message,
+            _dstEid: action.dstEid,
+            _message: action.message,
             _options: options,
             _payInLzToken: false
         });
-        return LzSendParams({dstEid: _dstEid, gasLimit: _gasLimit, fee: fee, options: options});
+        return LzSendParams({gasLimit: _gasLimit, fee: fee, options: options});
     }
 
     /// @notice Relay actions to another chain. Requires the sender to be authorized and the peer OApp to be set.
     /// @param _callId The unique identifier for the relayed actions, such as a proposal ID.
     /// @param _actions The actions to relay to the destination chain, including value, target and calldata.
     /// @param _allowFailureMap A bitmap of actions that are allowed to fail.
-    /// @param _params Additional Layer Zero params required to send a cross chain message, use the `quote` function to get these.
-    function relayActions(
+    /// @param _dstEid The LayerZero endpoint ID of the destination chain.
+    function queueRelayActions(
         uint256 _callId,
+        uint32 _dstEid,
         IDAO.Action[] memory _actions,
-        uint256 _allowFailureMap,
-        LzSendParams memory _params
-    ) external payable auth(XCHAIN_ACTION_RELAYER_ID) returns (MessagingReceipt memory receipt) {
+        uint256 _allowFailureMap
+    ) public auth(XCHAIN_ACTION_RELAYER_ID) {
         bytes memory message = abi.encode(_callId, _actions, _allowFailureMap);
+        _queueRelayActions(_callId, _dstEid, message);
+    }
+
+    function _queueRelayActions(uint256 _callId, uint32 _dstEid, bytes memory _message) internal {
+        require(actionsMap[_callId].message.length == 0, "ActionRelay: already queued");
```

Comment:
You can overwrite actions with a message.length of zero

Reply:
Changed it to use a boolean to mark the element as `queued`